### ISSUE TITLE
Cursors not removed when register out variable for insert statement

### ISSUE THREAD

**Issue by StigEkstrom** [vrogier/ocilib#51]
We get the exception "ORA-01000, maximum open cursors exceeded is thrown in the log" after a number of INSERT with RETURNING xxx INTO. We are using the c++ interface in ocilib 4.2.1.
I do not know if it is our uasage of ocilib that is wrong oe if it is a bug in ocilib.

The code we use look something like this:

std::string sql = "INSERT INTO xxx (y0, y1, y2, y3, ... y14) VALUES (Sequence.nextval, :1, :2,:3, ... :14) RETURNING zz INTO :15);
Statement m_statement(connection);
m_statement.Prepare(sql);
m_statement.Register<big_uint>(15);
m_statement.Bind(1, ...);
...
m_statement.Bind((15, ...);
m_statement.ExecutePrepared();
ResultSet rs = m_statement.GetResultset();
if (rs++)
{
   ret64 = rs.Get<big_uint>(1); 
}
int affected = m_statement.GetAffectedRows();
connection.Commit();

**Comment by vrogier** [comment]
Hi,

Can you provide a working sample code reproducing the issue I use for
investigation ?
Variables scope is quite important in such case.

Thanks.

Regards,

Vincent

**Comment by vrogier** [comment]
Hi,

I canot reproduce the issue.
Tried a large loop to reach max cursor limit in case cursor handle would be leaked but it is not the case. 

Can you provide some code reproducing the issue ?

Thanks

```C
        con.Open("DB", "usr", "pwd", Environment::SessionDefault);

        for (i = 0; i < 100000; i++)
        {
            std::string sql = "INSERT INTO test_stmt(value) VALUES (1) RETURNING value into :1";
            Statement st(con);

            st.Prepare(sql);
            st.Register<int>(":1");
            st.ExecutePrepared();
            Resultset rs = st.GetResultset();
            if (rs++)
            {
                int value = rs.Get<int>(1);
            }
            std:: cout << i << " - " << st.GetAffectedRows() << std::endl;
        }
```

regards,

Vincent

**Comment by vrogier** [comment]
Hi,

Any news ?

Regards,

Vincent

**Comment by FroderFlisk** [comment]
I'm not promising an update next week but I think I can say that there will not be one earlier than than that at least.

**Comment by StigEkstrom** [comment]
I am trying to create a working example.
Regards,
Stig

**Comment by StigEkstrom** [comment]
Hi,

Attached is an example where I get some cursors left opened even though I close the connection for each write/remove.
This example does not generate the same amount of opened cursor as our live system, but there are cursors left open after closing the connection.
When you terminate the program all opened cursors are removed, you have to check for open cursors while the program waits for user input.
To see if there are any cursors allocated I use the following SQL statements:

SELECT a.value,
  s.username,
  s.sid,
  s.serial#
FROM v$sesstat a,
  v$statname b,
  v$session s
WHERE a.statistic# = b.statistic#
AND s.sid          =a.sid
AND b.name         = 'opened cursors current'
AND s.username    IS NOT NULL
AND a.value > 10
ORDER BY 1 DESC;

SELECT count (*), sid ,   sql_text,   USER_NAME
FROM v$open_cursor
WHERE sid = 7
group by sid ,   sql_text,   USER_NAME
order by 1 desc;

sid in the second SELECT shall be the one from the result of the first SELECT.

Regards,
Stig

**Comment by StigEkstrom** [comment]
Here is the SQL statement to create the table:

  CREATE TABLE "DPOWNERA"."ASDBC_REGISTRYDATA_TEST"
   (           "REGDA_ID" NUMBER NOT NULL ENABLE,
                "EUNIT_NAME" VARCHAR2(20 BYTE) NOT NULL ENABLE,
                "REGDA_PARAMETERNAME" VARCHAR2(255 BYTE),
                "REGDA_PARAMETERVALUE" VARCHAR2(255 BYTE),
                "REGDA_PARAMETERTYPE" VARCHAR2(30 BYTE),
                "REGDA_INSTANCENAME" VARCHAR2(20 BYTE),
                "REGDA_SECURITY" VARCHAR2(20 BYTE),
                "REGDA_KEYALIAS" VARCHAR2(30 BYTE),
                "REGDA_VISIBILITY" VARCHAR2(20 BYTE),
                "REGDA_DESCRIPTION" VARCHAR2(2000 BYTE),
                "REGDA_UPDATEDWHEN" VARCHAR2(20 BYTE),
                "REGDA_DEPENDSON" VARCHAR2(255 BYTE),
                "REGDA_MINMAX" VARCHAR2(80 BYTE),
                "REGDA_DEFAULTVALUE" VARCHAR2(255 BYTE),
                "REGDA_TYPICALVALUE" VARCHAR2(255 BYTE)
   ) SEGMENT CREATION IMMEDIATE
  PCTFREE 10 PCTUSED 40 INITRANS 1 MAXTRANS 255
 NOCOMPRESS LOGGING
  STORAGE(INITIAL 131072 NEXT 131072 MINEXTENTS 1 MAXEXTENTS 2147483645
  PCTINCREASE 0 FREELISTS 1 FREELIST GROUPS 1
  BUFFER_POOL DEFAULT FLASH_CACHE DEFAULT CELL_FLASH_CACHE DEFAULT)
 TABLESPACE "DP_DATA01" ;

/Stig

From: Ekstrom Stig
Sent: den 27 juni 2016 08:12
To: 'vrogier/ocilib'; vrogier/ocilib
Cc: Author
Subject: RE: [vrogier/ocilib] Cursors not removed when register out variable for insert statement (#51)

Hi,

Attached is an example where I get some cursors left opened even though I close the connection for each write/remove.
This example does not generate the same amount of opened cursor as our live system, but there are cursors left open after closing the connection.
When you terminate the program all opened cursors are removed, you have to check for open cursors while the program waits for user input.
To see if there are any cursors allocated I use the following SQL statements:

SELECT a.value,
  s.username,
  s.sid,
  s.serial#
FROM v$sesstat a,
  v$statname b,
  v$session s
WHERE a.statistic# = b.statistic#
AND s.sid          =a.sid
AND b.name         = 'opened cursors current'
AND s.username    IS NOT NULL
AND a.value > 10
ORDER BY 1 DESC;

SELECT count (*), sid ,   sql_text,   USER_NAME
FROM v$open_cursor
WHERE sid = 7
group by sid ,   sql_text,   USER_NAME
order by 1 desc;

sid in the second SELECT shall be the one from the result of the first SELECT.

Regards,
Stig

**Comment by vrogier** [comment]
Hi,

There is no attached code source.

Vincent

**Comment by StigEkstrom** [comment]
Hi,
Trying again.
/Stig

**Comment by vrogier** [comment]
Hi,

Still no source code in uploaded in this issue.

Vincent

**Comment by StigEkstrom** [comment]
Here is the source code in a zip-file renamed to end with .txt as .zip was not accepted.
[testReturnInto.zip.txt](https://github.com/vrogier/ocilib/files/336329/testReturnInto.zip.txt)

**Comment by vrogier** [comment]
Hi,

With you test app and multiple run i can reproduce the issue if disable Oraclie client statement cache using:

 `   pool.SetStatementCacheSize(0);`

Two runs allow to reach the cursor limit on my oracle box (about 300 cursors)

Good news: I can reproduce it with the C++ test app.
Bad news: I cannot reproduce yet using another C app (different app)
I double checked the C++ API and all release calls are performed.

I'm still investigating the issue.

Regards, 

Vincent

**Comment by vrogier** [comment]
Hi,

I found the issue that is occurs ONLY when statement cache is not enabled and will fix it asap.
when statement caching is enabled on pools (default on all version >= 10.1), the issue dos not occurs.

Thus:
* If your real code is disabling statement cache on the pool by setting it to 0 ?
* What is your oracle client and server versions ?
* What is your OCILIB version ?

Regards,

Vincent

**Comment by vrogier** [comment]
In fact, OCILIB releases OCI statement using OCIStmtRelease().
If a statement had executed a "returning into" clause, OCILIB passes a special paremeter value to OCIStmtRelease() for forcing OCI to clear it it from the its cache as when reusing he statement with another "returning into" clause, OCI do not behave weel and forget to notify of the resultset.
But OCIStmtRelease() fails if statement cache is not activated.
I've just realized that OCILIB does activate the statement cache for coonnection pools but not for sessions pools !! Thus, when releasing such statements, OCIStmtRelease fails and there is a cursor leak.

I will commit a fix for that asap and it will be included in the upcoming 4.3.0 release!

Regards,

Vincent

**Comment by FroderFlisk** [comment]
Thanks for fixing this!
Do you have a plan for when 4.3.0 would be released?
